### ISSUE TITLE
[1.3.2.1] Error: No os unibuild for null!

### ISSUE THREAD

**Issue by wojtkowiak** [meteor/meteor#6843]
#### Repro is here https://github.com/wojtkowiak/repro2 

I am having a really strange issue. I have already tried to fix it for few hours but I am helpless.

I have a package which registers a linter.
I had this package included in the project in the packages dir inside the app. It worked without any problems but on one of my colleagues workstations the project does not run. We have reinstalled Meteor completely and cleaned everything we could with no effect. I have published the package to atmosphere hoping it would solve the issues but also without any effect.
The workstation is **OSX 64bit 10.11.4**.
Tested on Ubuntu and Windows and it runs fine.

This is the full error I get when the app is in the linting stage:
```
/Users/phauza/.meteor/packages/omega_android-manifest-modifier/.0.0.1.147jumn++os+web.browser+web.cordova/plugin.android-manifest-modifier.os/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:116
      throw error;
            ^
Error: No os unibuild for null!
    at Error (<anonymous>)
    at lintUnibuild (/tools/isobuild/compiler.js:304:11)
    at /tools/isobuild/compiler.js:218:30
    at Array.forEach (native)
    at Function._.each._.forEach (/Users/phauza/.meteor/packages/meteor-tool/.1.3.1.nwdv3u++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:79:11)
    at Object.lint (/tools/isobuild/compiler.js:211:5)
    at lintBundle (/tools/isobuild/bundler.js:2701:54)
    at /tools/isobuild/bundler.js:2576:25
    at /tools/utils/buildmessage.js:271:13
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2493:31)
    at /tools/runners/run-app.js:591:36
    at Function.run (/tools/tool-env/profile.js:489:12)
    at bundleApp (/tools/runners/run-app.js:581:34)
    at AppRunner._runOnce (/tools/runners/run-app.js:634:35)
    at AppRunner._fiber (/tools/runners/run-app.js:887:28)
    at /tools/runners/run-app.js:411:12
```

The error is thrown here in compiler.js
```javascript
const unibuild = _.findWhere(isopack.unibuilds, {arch: sourceArch.arch});
if (! unibuild) {
  throw Error(`No ${ sourceArch.arch } unibuild for ${ isopack.name }!`);
}
```
So from what I have understood reading the compiler.js code, it fails because it has expected that my app (when isopack.name is null it means it is the main app) will have the arch == 'os' while it doesn't.

I did not have time to dig through the whole isopack code but I have checked few things.
I have added some console logs around the place where the error is thrown and `isopack.unibuilds` (where the unibuilds for my app are) have only unibuilds for `os.osx.x86_64`, `web.browser` and `web.cordova` - which seems ok to me :smile:

My package is compiled to 'os' arch (not sure if I checked the right file):
```
cat ~/.meteor/packages/omega_android-manifest-modifier/~0.0.1/unipackage.json 
{
  "name": "omega:android-manifest-modifier",
  "summary": "Plugin that injects certain settings into cordova android manifest.",
  "version": "0.0.1",
  "isTest": false,
  "plugins": [
    {
      "name": "android-manifest-modifier",
      "arch": "os",
      "path": "plugin.android-manifest-modifier.os/program.json"
    }
  ],
  "unibuilds": [
    {
      "kind": "main",
      "arch": "os",
      "path": "os-legacy.json"
    },
    {
      "kind": "main",
      "arch": "web.browser",
      "path": "web.browser-legacy.json"
    },
    {
      "kind": "main",
      "arch": "web.cordova",
      "path": "web.cordova-legacy.json"
    }
  ],
  "format": "unipackage-pre2"
```

My best guess it that it takes the `os` arch from my package and expects my app will also have that arch.
What is weird is that when the package is downloaded from atmosphere it is theoretically downloaded in the correct build:
```
Downloading missing local versions of package omega:android-manifest-modifier@0.0.1 : [ 'os.osx.x86_64' ]
Doing HTTP request:  GET https://warehouse.meteor.com/builds/kPKPw2dwAJb7Pb4eL/1460971625032/KQdNzLKjfq/omega:android-manifest-modifier-0.0.1-os+web.browser+web.cordova.tgz
                                                                                   
Changes to your project's package version selections:
                                              
omega:android-manifest-modifier  added, version 0.0.1
```

`os` means any OS, right? So maybe the problem is that it is not satisfied with `os.osx.x86_64`?

Repro is here https://github.com/wojtkowiak/repro2 
It is mantra sample app blog with three packages added:
```
omega:android-manifest-modifier
nathantreid:css-modules
jshint
```

This issue only occurs on OSX.
Let me know if you need additional info.

**Comment by hwrod** [comment]
FWIW, I'm receiving the same exact error, same platform, and the included npm package did involve linting.. But even after removing it from package.json, the error remained. The error went away after running:

     meteor npm cache clean
     rm -rf node_modules

and came back after `meteor npm install`ing the package again, so looks like the error is directly caused by the package, but the meteor-promise/promise_server exception doesn't tell much about what was wrong with the package.

**Comment by hwrod** [comment]
@benjamn @wojtkowiak Resolution for me: This was fixed by removing the dburles:eslint package from my project.

What's likely is that since the npm module I was including had an .eslintrc file, the package was not only linting my Meteor project's JS **but also the npm package's files**. Removing the linter fixed the "No os unibuild for null!" and everything works as normal. Will add the linter back in at some point.

**Comment by wojtkowiak** [comment]
@hwrod thanks for you insight. 
In my case, I've removed `jshint` but still got the same error. My package registers a linter indeed so it is probably related more/less to this fact. Not sure however how it is all connected with each other, maybe @benjamn already has some clue.

**Comment by trajano** [comment]
This is triggered by   `Plugin.registerLinter` but after all the processing is done for a package.  From what I can tell from the text that appears, it seems to fail after compiling the project (though to be honest I don't think it should be compiling the project to begin with when we are linting)  That may be the root cause of the bug.

**Comment by trajano** [comment]
A very bad but working way of fixing this is to change https://github.com/meteor/meteor/blob/baa3f6a152af96ff1e7a7f9f0e00c86c9cd922ee/tools/isobuild/compiler.js#L302 to

    var unibuild = _.filter(isopack.unibuilds, function(build) { return build.arch.startsWith(sourceArch.arch) } )[0];

**Comment by abernix** [comment]
@wojtkowiak Is `omega:android-manifest-modifier` published on GitHub?

I was able to reproduce with your repo (thank you!), so marking this as confirmed.

**Comment by wojtkowiak** [comment]
@abernix it is not published. It is one of my really really old packages :smile: I just used it to inject permissions into AndroidManifest. It is of course not really a linter, it just reads a `.manifest` file in which the permissions are stored. I could probably just do what I want without registering a `linter` or `compiler` but back then I thought I could hook into the build system so it would be "nicer".
I have pushed the source code now if you want to take a look:
https://github.com/wojtkowiak/manifest_modifier

If you will create a file like `android.manifest` with:
```xml
<manifest>
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />
    <uses-permission android:name="android.permission.CHANGE_NETWORK_STATE" />
    <uses-permission android:name="android.permission.CHANGE_WIFI_STATE" />
    <uses-permission android:name="android.permission.WAKE_LOCK" />
    <uses-permission android:name="android.permission.CAMERA"/>
    <uses-permission android:name="android.permission.RECORD_AUDIO"/>
</manifest>
```
it will inject those into AndroidManifest and save the modified version in `cordova-build-override`

**Comment by abernix** [comment]
Linter or not, I think that's a reasonable use.  Having the package readily available just makes the debugging easier when someone looks at this.  I suppose if you want to 'un-publish' it you could just drop it into the `packages` directory for that reproduction repo. :wink:

**Comment by benjamn** [comment]
@trajano thanks for tracking that down! Your fix is not so different from what I believe to be the "correct" fix: https://github.com/meteor/meteor/commit/8253e5862d3dc3b6bc868aa5b57adeff09edbc5d

**Comment by trajano** [comment]
Cool when will the release be? :dancer:

**Comment by abernix** [comment]
@trajano There's no specific date.  Release frequency is trying to be sped up as much as possible, but follow the [milestone progress here](https://github.com/meteor/meteor/milestones/Release%201.3.3).  As always, if you want to be cutting edge, try running Meteor [from a checkout](https://github.com/meteor/meteor/blob/release-1.3.3/README.md#slow-start-for-developers), using the current `release-1.3.3` branch:
```sh
cd path/to/meteor
git fetch origin
git checkout -t origin/release-1.3.3
cd path/to/your/app
path/to/meteor/meteor run
```

And report back if the problem wasn't fixed for you.

**Comment by trajano** [comment]
@abernix will the checkout build work on Windows?  I haven't really tried.

**Comment by abernix** [comment]
@trajano Yup!  Just replace the final `meteor` in those instructions above with `meteor.bat` (I updated the instructions above).  Warning: It will most likely be slower running from checkout than the final build, but you can test new features and see how they work.

**Comment by tab00** [comment]
I've just received this error now after updating my app. I can no longer run the app. I hope the fix will be available soon.

**Comment by abernix** [comment]
@tab00 Are you using the `dburles:eslint` package?

**Comment by tab00** [comment]
> Are you using the dburles:eslint package?

No, but I use `east5th:check-checker` ("An eslint based check checker.").

Just to see if that was the cause I removed the package and my app is now working. So the problem is not as urgent as it was 20 minutes ago.

**Comment by abernix** [comment]
@tab00 Ok, good to know!  If you're able to run Meteor from checkout (as outlined above), it'd be nice to know if 8253e5862d3dc3b6bc868aa5b57adeff09edbc5d fixes it for you.

**Comment by benjamn** [comment]
This should be fixed if you do `meteor update --release 1.3.3-beta.0`.
